File: shopping cart.java
import java.util.HashMap;
import java.util.Map;

public class ShoppingCart {
    private Customer customer;
    private Map<Product, Integer> items;

    public ShoppingCart(Customer customer) {
        this.customer = customer;
        this.items = new HashMap<>();
    }

    public void addProduct(Product product, int quantity) {
        items.put(product, items.getOrDefault(product, 0) + quantity);
    }

    public void removeProduct(Product product) {
        items.remove(product);
    }

    public double getTotal() {
        double total = 0.0;
        for (Map.Entry<Product, Integer> entry : items.entrySet()) {
            total += entry.getKey().getPrice() * entry.getValue();
        }
        return total;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("ShoppingCart for ").append(customer.getName()).append("\n");
        for (Map.Entry<Product, Integer> entry : items.entrySet()) {
            sb.append(entry.getKey().getName()).append(" - ").append(entry.getValue()).append("\n");
        }
        sb.append("Total: $").append(getTotal());
        return sb.toString();
    }
}

File: shopping cart.py
class ShoppingCart:
    def __init__(self, customer):
        self.__customer = customer
        self.__items = {}  # Dictionary: Product → Quantity

    def add_product(self, product, quantity):
        if product in self.__items:
            self.__items[product] += quantity
        else:
            self.__items[product] = quantity

    def remove_product(self, product):
        if product in self.__items:
            del self.__items[product]

    def get_total(self):
        total = 0.0
        for product, quantity in self.__items.items():
            total += product.get_price() * quantity
        return total

    def __str__(self):
        cart_str = f"ShoppingCart for {self.__customer.get_name()}\n"
        for product, quantity in self.__items.items():
            cart_str += f"{product.get_name()} - {quantity}\n"
        cart_str += f"Total: ${self.get_total():.2f}"
        return cart_str
